File: agentscope-service/aistio/sdk/python/aistio/transport/registration.py
# Copyright 2024-2026 the original author or authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""v5 External Agent registration client."""
from __future__ import annotations

from dataclasses import dataclass
import json
from typing import Iterable
from urllib import error, request


@dataclass(frozen=True)
class RegisteredIdentity:
    """Stable Catalog identity required by every ASDP connection."""

    agent_id: str
    agent_key: str
    binding_id: str
    instance_id: str
    instance_key: str
    generation: int
    registration_credential: str


def register_external_agent(
    control_plane_http: str,
    *,
    bootstrap_token: str,
    registration_credential: str,
    tenant: str,
    namespace: str,
    agent_key: str,
    instance_key: str,
    routing_key: str,
    framework: str,
    sdk_version: str,
    capabilities: Iterable[str],
    timeout: float = 5.0,
) -> RegisteredIdentity:
    """Create or reclaim one External binding/instance through the v5 API."""

    body = json.dumps(
        {
            "tenant": tenant,
            "namespace": namespace,
            "agentKey": agent_key,
            "instanceKey": instance_key,
            "routingKey": routing_key,
            "framework": framework,
            "sdkVersion": sdk_version,
            "capacity": 1,
            "capabilities": list(capabilities),
        }
    ).encode("utf-8")
    headers = {"Content-Type": "application/json"}
    credential = registration_credential or bootstrap_token
    if credential:
        headers["Authorization"] = f"Bearer {credential}"
    req = request.Request(
        control_plane_http.rstrip("/") + "/api/v1/agent-registrations",
        data=body,
        headers=headers,
        method="POST",
    )
    try:
        with request.urlopen(req, timeout=timeout) as response:
            document = json.loads(response.read().decode("utf-8"))
    except error.HTTPError as exc:
        detail = exc.read().decode("utf-8", errors="replace")
        raise RuntimeError(f"agent registration returned HTTP {exc.code}: {detail}") from exc

    agent = document.get("agent") or {}
    binding = document.get("binding") or {}
    instance = document.get("instance") or {}
    issued = document.get("registrationCredential") or registration_credential
    identity = RegisteredIdentity(
        agent_id=str(agent.get("id") or ""),
        agent_key=agent_key,
        binding_id=str(binding.get("id") or ""),
        instance_id=str(instance.get("id") or ""),
        instance_key=instance_key,
        generation=int(instance.get("generation") or 0),
        registration_credential=str(issued or ""),
    )
    if (
        not identity.agent_id
        or not identity.binding_id
        or not identity.instance_id
        or identity.generation <= 0
        or not identity.registration_credential
    ):
        raise RuntimeError("agent registration response is missing stable identity or credential")
    return identity

File: agentscope-service/aistio/sdk/dsh/src/registration.ts
/*
 * Copyright 2024-2026 the original author or authors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { ControlPlaneHttpClient } from './control-plane.js'
import type { Logger } from './logger.js'

export interface RegistrationOptions {
    tenant: string
    controlHttp: string
    internalToken: string
    agentName: string
    namespace: string
    instanceId: string
    baseUrl: string
    runtime: string
    framework: string
    contractLevel: number
    capabilities: string[]
    heartbeatIntervalMs: number
    log: Logger
    client?: ControlPlaneHttpClient
}

/**
 * POST /api/v1/dataplanes/register plus periodic heartbeats.
 * Failures are swallowed and retried — registration must never disturb the agent.
 */
export class HttpSelfRegistration {
    private readonly http: ControlPlaneHttpClient
    private registered = false
    private registeredInstanceId: string | undefined
    private generation: number | undefined
    private timer: ReturnType<typeof setInterval> | undefined

    constructor(private readonly options: RegistrationOptions) {
        this.http =
            options.client ??
            new ControlPlaneHttpClient(options.controlHttp, options.internalToken)
    }

    start(): void {
        if (this.timer) {
            return
        }
        void this.tryRegister()
        this.timer = setInterval(() => {
            void this.heartbeatSafe()
        }, this.options.heartbeatIntervalMs)
        this.timer.unref?.()
    }

    async close(): Promise<void> {
        if (this.timer) {
            clearInterval(this.timer)
            this.timer = undefined
        }
        if (!this.registered) {
            return
        }
        try {
            const instanceId = this.requireRegisteredInstanceId()
            await this.http.send('DELETE', `/api/v1/dataplanes/${instanceId}`, {
                generation: this.generation,
            })
            this.options.log.info(`unregistered instance ${instanceId}`)
        } catch (error) {
            this.options.log.debug('unregister failed', error)
        } finally {
            this.registered = false
            this.registeredInstanceId = undefined
            this.generation = undefined
        }
    }

    private async heartbeatSafe(): Promise<void> {
        try {
            if (!this.registered) {
                await this.tryRegister()
                return
            }
            const instanceId = this.requireRegisteredInstanceId()
            const response = await this.http.send(
                'POST',
                `/api/v1/dataplanes/${instanceId}/heartbeat`,
                { generation: this.generation },
            )
            if (response.status === 404) {
                this.clearRegistration()
                await this.tryRegister()
            }
        } catch (error) {
            this.options.log.debug('heartbeat failed; will re-register', error)
            this.clearRegistration()
            try {
                await this.tryRegister()
            } catch {
                // swallowed
            }
        }
    }

    private async tryRegister(): Promise<void> {
        const body = {
            tenant: this.options.tenant,
            agentName: this.options.agentName,
            namespace: this.options.namespace,
            instanceId: this.options.instanceId,
            baseUrl: this.options.baseUrl.replace(/\/$/, ''),
            runtime: this.options.runtime,
            framework: this.options.framework,
            contractLevel: this.options.contractLevel,
            capabilities: this.options.capabilities,
            source: 'self-register',
        }
        try {
            const response = await this.http.send('POST', '/api/v1/dataplanes/register', body)
            if (response.status >= 200 && response.status < 300) {
                const registration = parseRegistrationResponse(response.json)
                this.registered = true
                this.registeredInstanceId = registration.instanceId
                this.generation = registration.generation
                this.options.log.info(
                    `registered ${registration.instanceId} (key=${this.options.instanceId}) at ${body.baseUrl} with ${this.options.controlHttp}`,
                )
            } else {
                this.clearRegistration()
                this.options.log.warn(`register returned HTTP ${response.status}`)
            }
        } catch (error) {
            this.clearRegistration()
            this.options.log.warn(
                `register failed (will retry): ${error instanceof Error ? error.message : String(error)}`,
            )
        }
    }

    private requireRegisteredInstanceId(): string {
        if (!this.registeredInstanceId) {
            throw new Error('control plane registration identity is missing')
        }
        return this.registeredInstanceId
    }

    private clearRegistration(): void {
        this.registered = false
        this.registeredInstanceId = undefined
        this.generation = undefined
    }
}

function parseRegistrationResponse(value: unknown): { instanceId: string; generation: number } {
    if (!value || typeof value !== 'object') {
        throw new Error('registration response must be a JSON object')
    }
    const instanceId = (value as Record<string, unknown>).instanceId
    const generation = (value as Record<string, unknown>).generation
    if (typeof instanceId !== 'string' || instanceId.length === 0) {
        throw new Error('registration response is missing instanceId')
    }
    if (typeof generation !== 'number' || !Number.isSafeInteger(generation) || generation < 1) {
        throw new Error('registration response is missing generation')
    }
    return { instanceId, generation }
}
